node/core/candidate-validation/src/lib.rs

```diff
@@ -0,0 +1,913 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The Candidate Validation subsystem.
+//!
+//! This handles incoming requests from other subsystems to validate candidates
+//! according to a validation function. This delegates validation to an underlying
+//! pool of processes used for execution of the Wasm.
+
+use polkadot_subsystem::{
+	Subsystem, SubsystemContext, SpawnedSubsystem, SubsystemResult,
+	FromOverseer, OverseerSignal,
+};
+use polkadot_subsystem::messages::{
+	AllMessages, CandidateValidationMessage, RuntimeApiMessage, ValidationFailed, RuntimeApiRequest,
+	RuntimeApiError,
+};
+use polkadot_node_primitives::{ValidationResult, ValidationOutputs};
+use polkadot_primitives::v1::{
+	ValidationCode, OmittedValidationData, PoV, CandidateDescriptor, LocalValidationData,
+	GlobalValidationData, OccupiedCoreAssumption, Hash, validation_data_hash,
+};
+use polkadot_parachain::wasm_executor::{self, ValidationPool, ExecutionMode};
+use polkadot_parachain::primitives::{ValidationResult as WasmValidationResult, ValidationParams};
+
+use parity_scale_codec::Encode;
+use sp_core::traits::SpawnNamed;
+
+use futures::channel::oneshot;
+use futures::prelude::*;
+
+use std::sync::Arc;
+
+/// The candidate validation subsystem.
+pub struct CandidateValidationSubsystem<S>(S);
+
+impl<S> CandidateValidationSubsystem<S> {
+	/// Create a new `CandidateValidationSubsystem` with the given task spawner.
+	pub fn new(spawn: S) -> Self {
+		CandidateValidationSubsystem(spawn)
+	}
+}
+
+impl<S, C> Subsystem<C> for CandidateValidationSubsystem<S> where
+	C: SubsystemContext<Message = CandidateValidationMessage>,
+	S: SpawnNamed + Clone + 'static,
+{
+	fn start(self, ctx: C) -> SpawnedSubsystem {
+		SpawnedSubsystem {
+			name: "candidate-validation-subsystem",
+			future: run(ctx, self.0).map(|_| ()).boxed(),
+		}
+	}
+}
+
+async fn run(
+	mut ctx: impl SubsystemContext<Message = CandidateValidationMessage>,
+	spawn: impl SpawnNamed + Clone + 'static,
+)
+	-> SubsystemResult<()>
+{
+	let pool = ValidationPool::new();
+
+	loop {
+		match ctx.recv().await? {
+			FromOverseer::Signal(OverseerSignal::ActiveLeaves(_)) => {}
+			FromOverseer::Signal(OverseerSignal::BlockFinalized(_)) => {}
+			FromOverseer::Signal(OverseerSignal::Conclude) => return Ok(()),
+			FromOverseer::Communication { msg } => match msg {
+				CandidateValidationMessage::ValidateFromChainState(
+					descriptor,
+					pov,
+					response_sender,
+				) => {
+					let res = spawn_validate_from_chain_state(
+						&mut ctx,
+						Some(pool.clone()),
+						descriptor,
+						pov,
+						spawn.clone(),
+					).await;
+
+					match res {
+						Ok(x) => { let _ = response_sender.send(x); }
```

Comment:
Should we warn if sending failed rather than silently dropping the result

Reply:
Sure. We might get some spurious warnings but probably OK.